### PR TITLE
Anim frame rewriting

### PR DESCRIPTION
**Pull request by TeslaRus**
Skeletal model frame system rewriting;
Entity frame rewriting;
INFO:
- underwater switch behavior is a script problem (underwater switch has other offset move, not hardcoded 360);
- no more interpolations in move or change direction commands case;

TODO:
- resolve change direction command with change frame inside animation...
- fix climb: crtl + forward near high wall - fixed jump condition;
- fix SET_POS command condition - fixed;

### CODE REVIEW

File: src/anim_state_control.cpp [hunk 1]
```diff
@@ -87,30 +87,9 @@ void ent_set_turn_fast(entity_p ent, ss_animation_p ss_anim, int state)
     }
 }
 
-void ent_set_on_floor_after_climb(entity_p ent, ss_animation_p ss_anim, int state)
```
Comment (TeslaRus) [inline src/anim_state_control.cpp]: That function may be used in other places to join animations frames

File: src/resource.cpp [hunk 2]
```diff
@@ -1700,20 +1633,16 @@ void TR_GenSkeletalModel(struct skeletal_model_s *model, size_t model_id, struct
         bone_frame = model->animations->frames;
 
         model->animations->id = 0;
-        model->animations->next_anim = NULL;
+        model->animations->next_anim = model->animations->next_anim;
```
Comment (TeslaRus) [inline src/resource.cpp]: should be model->animations->next_anim = model;

File: src/resource.cpp [hunk 3]
```diff
@@ -1955,8 +1882,9 @@ void TR_GenSkeletalModel(struct skeletal_model_s *model, size_t model_id, struct
     {
         anim->state_change_count = 0;
         anim->state_change = NULL;
-
-        tr_animation = &tr->animations[tr_moveable->animation_index+i];
+        anim->next_anim = anim;
```
Comment (TeslaRus) [inline src/resource.cpp]: anim is cycled by default

File: src/script.cpp [hunk 4]
```diff
@@ -2202,99 +2202,6 @@ int lua_SetStateChangeRange(lua_State * lua)
 }
 
 
-int lua_GetAnimCommandTransform(lua_State * lua)
```
Comment (TeslaRus) [inline src/script.cpp]: no fields, no setup

File: src/entity.cpp [hunk 5]
```diff
@@ -656,213 +656,245 @@ void Entity_CheckCollisionCallbacks(entity_p ent)
...
-                case TR_ANIMCOMMAND_EMPTYHANDS:
-                    ///@FIXME: Behaviour is yet to be discovered.
-                    break;
-
-                case TR_ANIMCOMMAND_KILL:
-                    // This command executes ONLY at the end of animation.
-                    if(ss_anim->current_frame == af->frames_count - 1)
-                    {
-                        if(entity->character)
+            for(uint32_t i = 0; i < count; i++, pointer++)
```
Comment (TeslaRus) [inline src/entity.cpp]: to parse all commands in all available places I will add some hack to the cycle (I hope I will found solution without hack)

File: src/engine.cpp [hunk 6]
```diff
@@ -906,8 +906,9 @@ void ShowDebugInfo()
                 entity_p ent = World_GetPlayer();
                 if(ent && ent->character)
                 {
-                    GLText_OutTextXY(30.0f, y += dy, "curr_anim = %03d, next_anim = %03d, curr_st = %03d, next_st = %03d", ent->bf->animations.current_animation, ent->bf->animations.next_animation, ent->bf->animations.current_state, ent->bf->animations.next_state);
```
Comment (TeslaRus) [inline src/engine.cpp]: todo: restore next anim / frame printing

**Comment by vvs-** [comment]
The first more or less playable version. Still problems with doors not opening or closing sometimes. There is no sound for crumbled floor animation.

**Comment by TeslaRus** [comment]
the original anim command bytecode will be transformed to the readable command sequence (like trigger command sequence) - more readable code + easier to fix bugs / correct algorithms. The main progress is eliminating GetNextFrame function. Main problem: some commands should be executed in frame starting, some - on frame ending.

File: scripts/autoexec.lua [hunk 7]
```diff
@@ -238,14 +238,6 @@ setModelCollisionMap(0, 12, 6);
 setModelCollisionMap(0, 13, 10);            -- braces 3
 setModelCollisionMap(0, 14, 13);
 
-setAnimCommandTransform(0, 147,  0, 0x00);   -- roll animation smooth fix
```
Comment (TeslaRus) [inline scripts/autoexec.lua]: removed hack

File: src/anim_state_control.cpp [hunk 8]
```diff
@@ -156,24 +156,24 @@ void ent_to_on_water(entity_p ent, ss_animation_p ss_anim, int state)
...
+        ent->no_fix_all = 0x01;
         ss_anim->onEndFrame = NULL;
     }
 }
 
-void ent_to_edge_climb(entity_p ent, ss_animation_p ss_anim, int state)
+void ent_to_edge_climb(entity_p ent, ss_animation_p ss_anim)
 {
-    if(state == 0x02)
+    if(ss_anim->changing_next >= 0x01)
```
Comment (TeslaRus) [inline src/anim_state_control.cpp]: check this

File: src/skeletal_model.h [hunk 9]
```diff
@@ -65,18 +65,25 @@ typedef struct ss_bone_tag_s
     uint32_t                body_part;                                          // flag: BODY, LEFT_LEG_1, RIGHT_HAND_2, HEAD...
 }ss_bone_tag_t, *ss_bone_tag_p;
 
-
+// changing info:
```
Comment (TeslaRus) [inline src/skeletal_model.h]: Add special constants, instead "magic" numbers

File: src/resource.cpp [hunk 10]
```diff
@@ -1552,9 +1553,101 @@ void TR_GenRoomMesh(struct room_s *room, size_t room_index, struct anim_seq_s *a
...
+                        bf->bone_tags[k].offset[2] = t * anim->frames[j-1].bone_tags[k].offset[2] + lerp * anim->frames[j].bone_tags[k].offset[2];
+
+                        vec4_slerp(bf->bone_tags[k].qrotate, anim->frames[j-1].bone_tags[k].qrotate, anim->frames[j].bone_tags[k].qrotate, lerp);
+                    }
+                    bf++;
+                }
+            }
+
+            /*
+             * swap old and new animation bone brames
```
Comment (stohrendorf) [inline src/resource.cpp]: bone "brames" ;)

**Comment by Gh0stBlade** [comment]
I tried this a few days ago, looking good. Will this PR eventually fix Lara's speed issues? Still getting issues where Lara moves too far. Most notably, when pressing back on a standard platform, Lara goes over the edge.

File: src/resource.cpp (hunk 10 above)
Comment (TeslaRus) [inline src/resource.cpp]: fixed =)

**Comment by TeslaRus** [comment]
@Gh0stBlade  
> Most notably, when pressing back on a standard platform, Lara goes over the edge. 
Fixed in last commit: really frames_count >= max_frame...

**Pull request merged**